Minimize entry parsing

kamidon opened the pull request:
Prior to this commit, while searching for the most applicable entry
password-store entries were decrypted and parsed to ensure they were
valid. The entries were parsed in the order they were found on the
filesystem and all applicable entries would be decrypted and parsed,
which varied based on the contents of the password-store and the entry
to be found.

This is fine when the GPG key is cached and each entry can be
decrypted without user interaction. However, for security some people
have their GPG on a hardware token like a Yubikey setup so that they
have to touch a sensor on the toke for every cryptographic operation,
in which case it becomes inconvenient as each attempt to find an entry
requires a variable number of touches of the hardware token.

The implementation already assumes that names which contain more of
the information in the search key should be preferred so there is an
ordering of preference of applicable entries. If the decrypt and
parsing is removed from the initial identification of applicable
entries in the store then in most cases a single decrypt and parse of
the most preferred entry will suffice, improving the experience for
hardware token users that require interaction with the token.

This commit implements that strategy. It is in spirit a refactor of
the existing code. The core of the change is the function
auth-source-pass--applicable-entries, which generates an ordered list
of regular expression matchers for all possible names that could be in
the password-store for the entry to be found and then makes a pass
over the password-store entry names accumulating the matching entries
in a list after the regexp that matched. This implementation ensures
the password-store entry list still only has to be scanned once.

The existing auth-source-pass--find-match-unambiguous was modified to
use this new function to obtain candidate entries and then parse them
one by one until an entry containing the desired information is
located. When complete it now returns the parsed data of the entry
instead of the entry name so that the information can be used directly
to construct the auth-source response.

One additional test case was added to the test suite to verify that
only the minimal number of entries are parsed in common cases. The
auth-source-pass-only-return-entries-that-can-be-open test case had to
be re-implemented because the function it used eliminated as the
functionality is provided elsewhere. All the other fairly substantial
changes to the test suite are the result of mechanical changes that
were required to adapt to auth-source-pass--find-match returning the
data from a parsed password-store entry instead of the entry name.

kamidon commented:
Sorry for the silly follow-up commit. I forgot to run `make check` instead of `make test`.

DamienCassou commented:
You did an amazing job!! Thank you so much. I think you deserve your name in the list of authors at the beginning of the file. Care to add it there?

I will review your code and give you feedback.

DamienCassou commented on test/auth-source-pass-tests.el:
```diff
@@ -88,125 +94,202 @@ This function is intended to be set to `auth-source-debug`."
-  (auth-source-pass--with-store '(("foo"))
-    (should (equal (auth-source-pass--find-match "foo" nil nil)
-                   "foo"))))
+  (auth-source-pass--with-store
```
I don't really like the fact that tests became more complex (they were already too complex for my taste). What if you changed that kind of test to evaluate `--applicable-entries` instead?

kamidon replied:
I like this idea as I'm not a fan of how the tests changed either. I was aiming for minimal transformation to the existing tests in the first version, but since you suggested changing them I'll try it out.  :-)

DamienCassou replied:
I'm still unsatisfied with the result but I will merge anyway. Thank you for having a look.

DamienCassou commented:
There are functions that can be removed: e.g., `auth-source-pass--select-one-entry` and `auth-source-pass--entry-valid-p`.

kamidon commented:
Thanks for the detailed comments. From a quick skim they are great suggestions. I may not be able to get to addressing them for a couple of days, but I'll follow up in detail as soon as I can make some time.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+                  (auth-source-pass--name-port-user-possibilities n user port))
+                base-names)))
+
+(defun auth-source-pass--name-possibilities (name-components)
```
Sure, I like just `...-domains`. I'll make that change.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+Based on the supported pathname patterns for HOSTNAME, USER, &
+PORT, return a list of possible suffixes for matching entries in
+the password-store."
+  (let ((base-names (auth-source-pass--name-possibilities
```
Definitely makes sense given the previous change.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+components, from longest to shortest.  The shortest sequence
+returned contains two components under the assumption that no-one
+will want to associate a pass entry with an entire TLD."
+  (when (> (length name-components) 0)
```
Good catch. It did originally function as documented but then I realized that would require changing a bunch of tests. I didn't see much harm in just checking the top-level as well even though I don't think it will be used in practice so I changed it to match all levels. I'll fix the docstring.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+          (auth-source-pass--name-possibilities (cdr name-components)))))
+
+(defun auth-source-pass--name-port-user-possibilities (name user port)
+  "Return a list of possible path suffixes for NAME, USER, & PORT."
```
Sure.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+                       (auth-source-pass-entries)
+                       :initial-value possibilities-match-lists))))
+
+(defun auth-source-pass--generate-entry-possibilities (hostname user port)
```
I replaced `...-possibilities` with `...-suffixes`. 

I'm not sure about pushing the regular expressions all the way down into what is now `auth-source-pass--generate-entry-suffixes`. I'm not strongly opposed to it but I'm not sure the extra flexibility will actually be used and I think doing so reduces the readability of the name formats. Unless you already have a use case for it, I would prefer to keep it as is for now.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+          (mapcar (lambda (possibility)
+                    (list (format "\\(^\\|/\\)%s$" possibility)))
+                  (auth-source-pass--generate-entry-possibilities hostname user port))))
+    (mapcar #'cdr
```
Yeah, this works well but I agree it isn't very readable. I'll try to split it up into something that explains what is going on better.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+   (auth-source-pass--applicable-entries hostname user port)
+   :initial-value nil))
+
+(defun auth-source-pass--select-from-entry-set (entry-set user)
```
I'll make that substitution everywhere it makes sense.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
-        (mapconcat 'identity (cdr components) ".")
-        user
-        port)))))
+  (cl-reduce
```
I'll try it and see how it feels.

kamidon commented on auth-source-pass.el:
```diff
@@ -265,33 +235,105 @@ inside USER by giving priority to USER.  Same for PORT."
+   (lambda (result entry-set)
+     (or result
+         (pcase (length entry-set)
+           (0 (auth-source-pass--do-debug "no match found")
```
What information would you like to capture in the log, the possibilities that were actually found in the store, or all of the possibilities that were evaluated for whether they were in the store? What do you think about `--applicable-entries` just logging the flattened version of the result it would return or "no match found" if that is `nil`?

Note that changing this logging will also require changes to some of the tests.

DamienCassou replied:
What I would like is the list of possible suffixes, then the list of matching entries then the chosen one.

kamidon commented:
I've added commits for the simple changes requested. I still need to work on:
* making `auth-source-pass--applicable-entries` more readable 
* trying the simplification of `auth-source-pass--find-match-unambiguous`
* changing the logging based on your feedback

kamidon commented:
I've added a series of commits that addresses most of the review feedback. I think that the only significant piece of feedback that is not accounted for here is simplifying `auth-source-pass--find-match-unambiguous`, which I ran out of time to attempt. I probably won't have time to look at that for a little while again either. If the updates to respond to the rest of the feedback look good I'd like to propose merging this as-is and leave fixing `auth-source-pass--find-match-unambiguous` for a future commit.

kamidon commented:
Also, I'm honored to be included in the authors list. Thanks!

DamienCassou commented:
Awesome work, thank you very much. I rewrote the git history to simplify it. I feel that there is quite some code simplification to do but we will do that later. Feel free to send more PRs I love them.

kamidon commented:
Thanks for merging @DamienCassou. I agree that there is more that can be improved and will continue to work on possible improvements as I have time.
